src/uct/base/uct_md.c

```diff
@@ -384,17 +385,82 @@ ucs_status_t uct_rkey_release(uct_component_h component,
     return component->rkey_release(component, rkey_ob->rkey, rkey_ob->handle);
 }
 
+void uct_md_copy_v1_fields(uct_md_attr_t *dst, uct_md_attr_v2_t *src)
+{
+    dst->cap.max_alloc        = src->max_alloc;
+    dst->cap.max_reg          = src->max_reg;
+    dst->cap.flags            = src->flags;
+    dst->cap.reg_mem_types    = src->reg_mem_types;
+    dst->cap.cache_mem_types  = src->cache_mem_types;
+    dst->cap.detect_mem_types = src->detect_mem_types;
+    dst->cap.alloc_mem_types  = src->alloc_mem_types;
+    dst->cap.access_mem_types = src->access_mem_types;
+    dst->reg_cost             = src->reg_cost;
+    dst->rkey_packed_size     = src->rkey_packed_size;
+
+    memcpy(&dst->local_cpus, &src->local_cpus, sizeof(src->local_cpus));
+    memcpy(&dst->component_name, &src->component_name, sizeof(src->component_name));
+}
+
+void uct_md_copy_v2_fields(uct_md_attr_v2_t *dst, uct_md_attr_v2_t *src)
+{
+    UCT_MD_ATTR_SET_V2_FIELD(dst, src, max_alloc, UCT_MD_ATTR_FIELD_MAX_ALLOC);
+    UCT_MD_ATTR_SET_V2_FIELD(dst, src, max_reg, UCT_MD_ATTR_FIELD_MAX_REG);
+    UCT_MD_ATTR_SET_V2_FIELD(dst, src, flags, UCT_MD_ATTR_FIELD_FLAGS);
+    UCT_MD_ATTR_SET_V2_FIELD(dst, src, reg_mem_types, UCT_MD_ATTR_FIELD_REG_MEM_TYPES);
+    UCT_MD_ATTR_SET_V2_FIELD(dst, src, cache_mem_types, UCT_MD_ATTR_FIELD_CACHE_MEM_TYPES);
+    UCT_MD_ATTR_SET_V2_FIELD(dst, src, detect_mem_types, UCT_MD_ATTR_FIELD_DETECT_MEM_TYPES);
+    UCT_MD_ATTR_SET_V2_FIELD(dst, src, alloc_mem_types, UCT_MD_ATTR_FIELD_ALLOC_MEM_TYPES);
+    UCT_MD_ATTR_SET_V2_FIELD(dst, src, access_mem_types, UCT_MD_ATTR_FIELD_ACCESS_MEM_TYPES);
+    UCT_MD_ATTR_SET_V2_FIELD(dst, src, dmabuf_mem_types, UCT_MD_ATTR_FIELD_DMABUF_MEM_TYPES);
+    UCT_MD_ATTR_SET_V2_FIELD(dst, src, reg_cost, UCT_MD_ATTR_FIELD_REG_COST);
+    UCT_MD_ATTR_SET_V2_FIELD(dst, src, rkey_packed_size, UCT_MD_ATTR_FIELD_RKEY_PACKED_SIZE);
+    UCT_MD_ATTR_COPY_V2_FIELD(dst, src, local_cpus, UCT_MD_ATTR_FIELD_LOCAL_CPUS);
+    UCT_MD_ATTR_COPY_V2_FIELD(dst, src, component_name, UCT_MD_ATTR_FIELD_COMPONENT_NAME);
+}
+
 ucs_status_t uct_md_query(uct_md_h md, uct_md_attr_t *md_attr)
 {
     ucs_status_t status;
+    uct_md_attr_v2_t tmp_md_attr;
 
-    status = md->ops->query(md, md_attr);
+    memset(&tmp_md_attr, 0, sizeof(tmp_md_attr));
+
+    tmp_md_attr.field_mask = -1;
+
+    status = md->ops->query(md, &tmp_md_attr);
     if (status != UCS_OK) {
         return status;
     }
 
     /* Component name + data */
-    memcpy(md_attr->component_name, md->component->name, UCT_COMPONENT_NAME_MAX);
+    memcpy(tmp_md_attr.component_name, md->component->name, sizeof(md_attr->component_name));
+
+    uct_md_copy_v1_fields(md_attr, &tmp_md_attr);
+
+    return UCS_OK;
+}
+
+ucs_status_t uct_md_query_v2(uct_md_h md, uct_md_attr_v2_t *md_attr)
```

Comment:
there are duplications between `uct_md_query` and `uct_md_query_v2`.
pls, add common function which initializes `uct_md_attr_v2_t tmp_md_attr`, invokes `md->ops->query()` and initializes `tmp_md_attr`'s component name from `md`'s `component_name`.
and then this function could be re-used in `uct_md_query` and `uct_md_query_v2`.